libc/harness/test_exec.c

```diff
@@ -0,0 +1,201 @@
+
+		if (execve("/bin/test_exec", arg, environ) == -1) {
+			fprintf(stderr, "execve function failed\n");
+
+			exit(EXIT_FAILURE);
+		}
+	}
+}
+
+
+static void test_exec_execvePath(void)
+{
+	char *const arg[] = { "to_exec", NULL };
+
+	setenv("PATH", "/bin:/sbin:/usr/bin:/usr/sbin", 1);
+
+	if (execve("test_exec", arg, environ) == -1) {
+		fprintf(stderr, "execve function failed\n");
+
+		exit(EXIT_FAILURE);
+	}
+}
+
+
+static void test_exec_execvpeEnv(bool changeEnv)
+{
+	char *const arg[] = { "/bin/to_exec", NULL };
+	if (changeEnv) {
+		char *const env[] = { "TEST1=exec_value", NULL };
+		setenv("TEST1", "invalid_value", 1);
+		setenv("TEST2", "should_dissapear", 1);
+
+		if (execvpe("/bin/test_exec", arg, env) == -1) {
+			fprintf(stderr, "execvpe function failed\n");
+
+			exit(EXIT_FAILURE);
+		}
+	}
+	else {
+		setenv("TEST1", "unchanged_value", 1);
+
+		if (execvpe("/bin/test_exec", arg, environ) == -1) {
+			fprintf(stderr, "execvpe function failed\n");
+
+			exit(EXIT_FAILURE);
+		}
+	}
+}
+
+
+static void test_exec_execvpePath(void)
+{
+	char *const arg[] = { "to_exec", NULL };
+
+	setenv("PATH", "/bin:/sbin:/usr/bin:/usr/sbin", 1);
+
+	if (execvpe("test_exec", arg, environ) == -1) {
+		fprintf(stderr, "execve function failed\n");
+
+		exit(EXIT_FAILURE);
+	}
+}
+
+
+static void test_exec_execvpEnv(void)
+{
+	char *const arg[] = { "/bin/to_exec", NULL };
+
+	setenv("TEST1", "unchanged_value", 1);
+
+	if (execvp("/bin/test_exec", arg) == -1) {
+		fprintf(stderr, "execvpe function failed\n");
+
+		exit(EXIT_FAILURE);
+	}
+}
+
+
+static void test_exec_execvpPath(void)
+{
+	char *const arg[] = { "to_exec", NULL };
+
+	setenv("PATH", "/bin:/sbin:/usr/bin:/usr/sbin", 1);
+
+	if (execvp("test_exec", arg) == -1) {
+		fprintf(stderr, "execve function failed\n");
+
+		exit(EXIT_FAILURE);
+	}
+}
+
+
+int main(int argc, char *argv[], char const *envp[])
+{
+	bool changeEnv;
+
+	if (!strcmp(basename(argv[0]), "to_exec")) {
+		int i;
+		printf("argc = %d\n", argc);
+
+		for (i = 0; argv[i] != NULL; i++) {
+			printf("argv[%d] = %s\n", i, argv[i]);
+		}
+
+		for (i = 0; environ[i] != NULL; i++) {
+			printf("environ[%d] = %s\n", i, environ[i]);
+		}
+	}
+	else {
+		if ((argc == 1) || (atoi(argv[1]) > 10) || (atoi(argv[1]) < 0)) {
+			fprintf(stderr, "Please specify test case number (1 to 10)\n");
+			return 1;
+		}
+
+		switch (atoi(argv[1])) {
+			case 1:
+				changeEnv = true;
+				test_exec_execveEnv(changeEnv);
+				break;
+
+			case 2:
+				changeEnv = false;
+				test_exec_execveEnv(changeEnv);
+				break;
+
+			case 3:
+				changeEnv = true;
+				test_exec_execvePath();
+				break;
+
+			case 4:
+				changeEnv = true;
+				test_exec_execvpeEnv(changeEnv);
+				break;
+
+			case 5:
+				changeEnv = false;
+				test_exec_execvpeEnv(changeEnv);
+				break;
+
+			case 6:
+				test_exec_execvpePath();
+				break;
+
+			case 7:
+				test_exec_execvpEnv();
+				break;
+
+			case 8:
+				test_exec_execvpPath();
+				break;
```

Comment:
I would also add here some tests for:
 - `execl`, `execlp`, `execle`
 - passing non-trivial arguments (more than one :))
 - (in the future) execs with interpreter (so a script with a she-bang)
 - intentionally failing cases (and verify the `errno`) - eg. missing file, trying to exec directory, ENOMEM would be probably hard to get :)